code/server/sv_http.c

```diff
@@ -0,0 +1,1066 @@
+	const char *baseGame = Cvar_VariableString("com_basegame");
+	const int baseGameSize = (int)strlen(baseGame);
+	const char *basename;
+	char pk3Path[MAX_QPATH];
+	char *dispositionBuf;
+	int i;
+
+	if (!referencedPaks || !*referencedPaks) {
+		HTTP_ResponseSetText(response, "Server doesn't have any pk3 referenced\n");
+		return qtrue;
+	}
+
+	COM_StripExtension(request->path + 5, pk3Path, sizeof(pk3Path));
+	if (Q_strncmp(baseGame, pk3Path, baseGameSize) != 0) {
+		HTTP_ResponseSetText(response, "Invalid pk3 file request\n");
+		return qtrue;
+	}
+
+	basename = pk3Path + baseGameSize;
+
+	if (*basename != '/') {
+		HTTP_ResponseSetText(response, "Invalid request path\n");
+		return qtrue;
+	}
+	++basename;
+
+#ifndef STANDALONE
+	if (FS_idPak(pk3Path, BASEGAME, NUM_ID_PAKS)
+#ifdef MISSIONPACK
+		|| FS_idPak(pk3Path, BASETA, NUM_TA_PAKS)
+#endif
+	) {
+		HTTP_ResponseSetText(response, "Invalid request path\n");
+		return qtrue;
+	}
+#endif
+
+	Cmd_TokenizeString(referencedPaks);
+
+	for (i = 0; i < Cmd_Argc(); i++) {
+		if (!strcmp(pk3Path, Cmd_Argv(i))) {
+			// attachment; filename=""
+			const int bufSize = sizeof(pk3Path) + 24;
+			dispositionBuf = (char *)malloc(bufSize);
+			Com_sprintf(dispositionBuf, bufSize, "attachment; filename=\"%s.pk3\"", basename);
+			HTTP_SetHeader(&response->proto, HTTP_HEADER_CONTENT_TYPE, HTTP_MIMETYPE_APPLICATION_OCTET_STREAM, qfalse);
+			HTTP_SetHeader(&response->proto, HTTP_HEADER_CONTENT_DISPOSITION, dispositionBuf, qtrue);
+			HTTP_ResponseSetPk3File(response, va("%s.pk3", pk3Path));
+			return qtrue;
+		}
+	}
+	HTTP_ResponseSetText(response, "Given pk3 is not referenced\n");
+	return qtrue;
+}
+
+qboolean SV_HTTPServerInit(void) {
+	int port = sv_httpServerPort->integer;
+	struct sockaddr_in sin;
+	int t = 1;
+	int i;
+
+	if (port <= 0) {
+		Com_Printf("Built-in http server is disabled\n");
+		return qtrue;
+	}
+
+	memset(httpClients, 0, sizeof(httpClients));
+	for (i = 0; i < HTTP_MAX_CLIENTS; ++i) {
+		httpClients[i].socket = INVALID_SOCKET;
+	}
+
+	memset(httpEndpoints, 0, sizeof(httpEndpoints));
+
+	httpSocket = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
+	if (httpSocket == INVALID_SOCKET) {
+		return qfalse;
+	}
+	memset(&sin, 0, sizeof(sin));
+	sin.sin_family = AF_INET;
+	sin.sin_addr.s_addr = INADDR_ANY;
+	sin.sin_port = htons(port);
+
+	FD_ZERO(&httpReadFDSet);
+	FD_ZERO(&httpWriteFDSet);
+
+#ifdef _WIN32
+	if (setsockopt(httpSocket, SOL_SOCKET, SO_REUSEADDR, (char *)&t, sizeof(t)) != 0) {
+#else
+	if (setsockopt(httpSocket, SOL_SOCKET, SO_REUSEADDR, &t, sizeof(t)) != 0) {
+#endif
+		closesocket(httpSocket);
+		httpSocket = INVALID_SOCKET;
+		return qfalse;
+	}
+
+	if (bind(httpSocket, (struct sockaddr *)&sin, sizeof(sin)) < 0) {
+		closesocket(httpSocket);
+		httpSocket = INVALID_SOCKET;
+		return qfalse;
+	}
+
+	if (listen(httpSocket, 5) < 0) {
+		closesocket(httpSocket);
+		httpSocket = INVALID_SOCKET;
+		return qfalse;
```

Comment:
I see a couple inconsistencies with style on new lines before returns. I personally like a new line before a return, but as long as it consistent, it doesn't matter which way you do it.